# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import datetime
import seaborn as sns

path="/Users/atousachaharsoughshirazi/desktop/Bootcamp/Unit_6/data"
os.chdir(path) 
os.getcwd()

'/Users/atousachaharsoughshirazi/Desktop/Bootcamp/Unit_6/data'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df=pd.read_csv('step3_output.csv')
df.drop(df.columns[0], axis=1, inplace=True)
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
 # select object type columns
dfo=df['state'].values
df = pd.concat([df.drop(columns='state', axis=1), pd.get_dummies(dfo)], axis=1)
#df = pd.concat([df.drop([df.state], axis=1), pd.get_dummies(df.state)], axis=1)
df.head(-5)

Name  summit_elev  vertical_drop  base_elev  trams  \
0          Alyeska Resort         3939           2500        250      1   
1     Eaglecrest Ski Area         2600           1540       1200      0   
2        Hilltop Ski Area         2090            294       1796      0   
3        Arizona Snowbowl        11500           2300       9200      0   
4     Sunrise Park Resort        11100           1800       9200      0   
..                    ...          ...            ...        ...    ...   
320     Whitecap Mountain         1750            400       1295      0   
321       Wilmot Mountain         1030            230        800      0   
322  Grand Targhee Resort         9920           2270       7851      0   
323         Hogadon Basin         8000            640       7400      0   
324          Jackson Hole        10450           4139       6311      3   

     fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0          0.0          0          2     2       0  ...             0   
1          0.0          0          0     0       0  ...             0   
2          0.0          0          0     0       1  ...             0   
3          0.0          1          0     2       2  ...             0   
4          0.0          0          1     2       3  ...             0   
..         ...        ...        ...   ...     ...  ...           ...   
320        0.0          0          0     1       0  ...             0   
321        0.0          0          0     3       2  ...             0   
322        0.0          0          2     2       0  ...             0   
323        0.0          0          0     0       0  ...             0   
324        0.0          0          4     4       1  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
320             0          0     0        0         0           0   
321             0          0     0        0         0           0   
322             0          0     0        0         0           0   
323             0          0     0        0         0           0   
324             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
320              0          1        0  
321              0          1        0  
322              0          0        1  
323              0          0        1  
324              0          0        1  

[325 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
X_train

array([[-0.15866985, -0.91441494,  0.04048329, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.25518791,  1.47351277, -0.75311775, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.97639234, -0.9672452 , -0.87521021, ..., -0.11076976,
        -0.22573306, -0.15762208],
       ...,
       [-0.96968969, -0.99366033, -0.82701582, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [ 1.05853018,  2.61464637,  0.84372321, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [ 1.97330695, -0.05645153,  2.38273089, ..., -0.11076976,
        -0.22573306, -0.15762208]])

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
y_pred

array([ 58.18540601,  90.35288648,  78.56773023,  76.26694898,
        70.04331617,  76.01743726,  45.02866773,  46.83579664,
        93.27671461,  83.20249586,  51.13901929,  79.06577711,
        51.08970289,  71.48569898,  62.88706617,  92.86655836,
        37.14780836,  72.55503492,  61.35264234,  41.28843336,
       113.69663648,  36.79917554,  53.31870679,  49.28257398,
        48.48960523,  62.83335523,  77.29819898,  82.41733961,
        64.20249586,  66.41538648,  35.34312086,  61.39194898,
        79.84702711,  56.95249586,  75.22007398,  49.23569898,
        23.23960523,  29.13608961,  61.72837476,  37.55015211,
        47.93491773,  61.07358961,  85.78257398,  72.03940992,
        71.95591383,  57.49839429,  64.31431226, 164.16733961,
        38.11460523,  62.87632398,  68.46616773,  59.31187086,
        53.20249586,  45.77866773,  58.92515211,  42.20249586,
        87.28452711,  38.55210523, 118.57358961, 118.78648023,
        56.16440992,  56.01841383,  42.11558179,  79.75

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from . 
# You can plug y_test and y_pred into the functions to evaluate the model

In [9]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)


0.9278481935148055

In [10]:
mean_absolute_error(y_test, y_pred)

5.210508545284814

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [11]:
model.intercept_

64.12827710698181

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [12]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        2.083427e+01
vertical_drop      6.479283e+00
base_elev          1.906166e+01
trams              2.211129e+12
fastEight          2.173757e+11
fastSixes          2.573391e+12
fastQuads          8.680677e+12
quad               5.181824e+12
triple             6.393659e+12
double             7.167228e+12
surface            8.133143e+12
total_chairs       2.289798e+13
Runs               1.273438e+00
TerrainParks       2.373962e-01
LongestRun_mi      4.072266e-01
SkiableTerrain_ac  1.914062e-01
Snow Making_ac     9.147949e-01
daysOpenLastYear   2.170410e-01
yearsOpen          3.880615e-01
averageSnowfall    2.463379e-01
AdultWeekday       1.913843e+01
projectedDaysOpen  6.542969e-02
NightSkiing_ac     7.666016e-01
clusters           1.083984e+00
Alaska             7.803506e+11
Arizona            6.381271e+11
California         2.006985e+12
Colorado           2.050888e+12
Connecticut        1.004343e+12
Idaho              1.539074e+12
Illinois           8.996925e+11
Indiana            6.381271e+11
Iowa               7.803506e+11
Maine              1.339149e+12
Maryland           4.519113e+11
Massachusetts      1.475866e+12
Michigan           2.327756e+12
Minnesota          1.657154e+12
Missouri           6.381271e+11
Montana            1.539074e+12
Nevada             8.996925e+11
New Hampshire      1.765957e+12
New Jersey         6.381271e+11
New Mexico         1.339149e+12
New York           2.466553e+12
North Carolina     1.098508e+12
Ohio               1.004343e+12
Oregon             1.409387e+12
Pennsylvania       1.915192e+12
Rhode Island       4.519113e+11
South Dakota       6.381271e+11
Tennessee          4.519113e+11
Utah               1.599398e+12
Vermont            1.712601e+12
Virginia           8.996925e+11
Washington         1.409387e+12
West Virginia      8.996925e+11
Wisconsin          1.765957e+12
Wyoming            1.264527e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [13]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df

X2=X.iloc[:,:-35]
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[-1.58669848e-01, -9.14414942e-01,  4.04832918e-02, ...,
        -1.04716512e+00, -6.09611897e-01,  7.80868809e-02],
       [-2.55187912e-01,  1.47351277e+00, -7.53117745e-01, ...,
         5.20812283e-01, -6.09611897e-01,  7.80868809e-02],
       [-9.76392337e-01, -9.67245201e-01, -8.75210213e-01, ...,
        -9.90324405e-16, -3.31010103e-01,  7.80868809e-02],
       ...,
       [-9.69689694e-01, -9.93660331e-01, -8.27015818e-01, ...,
        -6.98725699e-01, -2.34571021e-01,  7.80868809e-02],
       [ 1.05853018e+00,  2.61464637e+00,  8.43723208e-01, ...,
         6.44428248e+00,  2.60502419e+00, -1.53245504e+00],
       [ 1.97330695e+00, -5.64515333e-02,  2.38273089e+00, ...,
         5.55656226e-01, -6.09611897e-01,  1.68862880e+00]])

In [15]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [16]:
y_pred=model.predict(X2_test)
y_pred

array([ 57.14824007,  85.58010879,  82.88339407,  78.39171672,
        66.94711561,  81.66153492,  38.08828682,  47.76616671,
        88.70466733,  85.2890039 ,  52.10555351,  80.68436424,
        52.87536858,  65.70582409,  63.5166317 ,  88.72619354,
        37.66303169,  77.89432436,  63.19930589,  43.46319216,
       109.27837179,  37.4525708 ,  54.01722104,  50.30562505,
        50.59311144,  61.05811115,  66.46053759,  75.84537121,
        63.46035301,  64.71039803,  38.73129417,  53.98677059,
        78.47889126,  57.34158217,  75.49295886,  54.20417401,
        25.39593932,  28.4003079 ,  55.15250765,  39.05312695,
        49.83224716,  63.53734671,  88.43372824,  66.80283044,
        68.87401887,  52.80661293,  64.43320602, 166.12549512,
        30.79666359,  64.77320011,  62.98568729,  60.56309853,
        53.8743828 ,  47.02102787,  60.98317755,  41.92378747,
        88.43774454,  31.38279673, 120.32821205, 115.59753733,
        57.89963234,  55.81727404,  39.8743546 ,  86.14

In [17]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)

0.9223038219830924

In [18]:
mean_absolute_error(y_test, y_pred)

5.543225362838364

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [19]:
model.intercept_

64.07822304615368

In [20]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])



Coefficient
summit_elev          13.955579
vertical_drop         4.949809
base_elev            10.725078
trams                 0.350678
fastEight             0.070302
fastSixes             0.384010
fastQuads             0.670655
quad                  1.530338
triple                1.426216
double                0.380747
surface               1.292115
total_chairs          0.667653
Runs                  1.525236
TerrainParks          0.242733
LongestRun_mi         0.242468
SkiableTerrain_ac     0.358463
Snow Making_ac        0.500205
daysOpenLastYear      0.960395
yearsOpen             0.286668
averageSnowfall       1.509100
AdultWeekday         19.940382
projectedDaysOpen     0.350776
NightSkiing_ac        0.471929
clusters              0.929127

#### Model 3

In [21]:
X2.drop(['summit_elev','base_elev'],axis=1, inplace=True)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 

y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 


In [22]:
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[-9.14414942e-01, -3.08939686e-01, -5.51317846e-02, ...,
        -1.04716512e+00, -6.09611897e-01,  7.80868809e-02],
       [ 1.47351277e+00, -3.08939686e-01, -5.51317846e-02, ...,
         5.20812283e-01, -6.09611897e-01,  7.80868809e-02],
       [-9.67245201e-01, -3.08939686e-01, -5.51317846e-02, ...,
        -9.90324405e-16, -3.31010103e-01,  7.80868809e-02],
       ...,
       [-9.93660331e-01, -3.08939686e-01, -5.51317846e-02, ...,
        -6.98725699e-01, -2.34571021e-01,  7.80868809e-02],
       [ 2.61464637e+00, -3.08939686e-01, -5.51317846e-02, ...,
         6.44428248e+00,  2.60502419e+00, -1.53245504e+00],
       [-5.64515333e-02, -3.08939686e-01, -5.51317846e-02, ...,
         5.55656226e-01, -6.09611897e-01,  1.68862880e+00]])

In [23]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [24]:
y_pred=model.predict(X2_test)
y_pred

array([ 57.74347611,  85.02743353,  84.12013482,  78.60197973,
        65.69963768,  82.66066285,  37.70117772,  47.31131701,
        88.36843442,  84.71239695,  51.89506586,  81.93620381,
        53.26719898,  65.26179036,  63.18308585,  88.27855506,
        40.72189668,  78.72567138,  65.42590651,  43.24586766,
       108.521336  ,  36.96977743,  52.87629418,  50.09356341,
        50.43344603,  61.61665157,  66.73224886,  72.61874563,
        64.29358849,  64.54470312,  40.05464517,  53.93993436,
        78.74829128,  56.61679198,  74.86238531,  54.17282425,
        25.17400755,  27.96507427,  54.92460117,  38.58368002,
        50.29893364,  64.12668174,  88.20541914,  66.30396697,
        69.112616  ,  52.33379233,  64.83139188, 166.55798936,
        30.65263115,  65.86620567,  62.53184169,  61.42393793,
        53.83112475,  47.18060971,  61.13189985,  42.62355622,
        87.30276148,  32.02555645, 119.07428623, 115.17353657,
        57.40982621,  56.296763  ,  39.4543597 ,  85.61

In [25]:
explained_variance_score(y_test, y_pred)

0.9239026961723045

In [26]:
mean_absolute_error(y_test, y_pred)

5.53109010511069

In [27]:
model.intercept_

64.07667128280998

In [28]:
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])

Coefficient
vertical_drop         1.291189
trams                 0.300407
fastEight             0.101771
fastSixes             0.373185
fastQuads             0.771395
quad                  1.578662
triple                1.380822
double                0.378014
surface               1.266572
total_chairs          0.717592
Runs                  1.665804
TerrainParks          0.149166
LongestRun_mi         0.232629
SkiableTerrain_ac     0.438417
Snow Making_ac        0.402843
daysOpenLastYear      0.829641
yearsOpen             0.310477
averageSnowfall       1.982708
AdultWeekday         19.893397
projectedDaysOpen     0.095407
NightSkiing_ac        0.606307
clusters              0.766623

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9276995091611128 | 5.215550672053289 |-|
| Model 2. | 0.9223038219830926| 5.543225362838355 |'state'|
| Model 3. | 0.9239026961723045 | 5.531090105110692 |'state','summit_elev','base_elev'|


Model Selection:

In [29]:
#Model 1